HTTP Sink: Send RFC5424 (syslog) messages over HTTP

kensimon opened the pull request:
This adds a sink type with similar semantics to Heroku's Logplex.

Notes:

- Default behavior is to buffer 1500 events for sending in case the HTTP drain
  is slow.  By default, if this buffer is exceeded, messages will be dropped.
  (Similar to logplex.)
- If multiple events have come in since the last request, they will be placed
  in a single request (separated with newlines and length prefixes per rfc5424)
  for the sake of speed.
- Failure to contact the HTTP server will retry up to 10 times with exponential
  backoff

kensimon commented:
Sample requests from the HTTP sink:

```
POST / HTTP/1.1
Host: localhost:8082
User-Agent: Go-http-client/1.1
Content-Length: 1717
Accept-Encoding: gzip

847 <24>1 2017-05-02T16:27:59-07:00 - ephemeral-577301787-1w711 - - - {"verb":"ADDED","event":{"metadata":{"name":"ephemeral-577301787-1w711.14baedca39165dfb","namespace":"kube-system","selfLink":"/api/v1/namespaces/kube-system/events/ephemeral-577301787-1w711.14baedca39165dfb","uid":"fa79f792-2f8e-11e7-956e-06d49b68d344","resourceVersion":"960017","creationTimestamp":"2017-05-02T23:27:59Z"},"involvedObject":{"kind":"Pod","namespace":"kube-system","name":"ephemeral-577301787-1w711","uid":"fa78e002-2f8e-11e7-956e-06d49b68d344","apiVersion":"v1","resourceVersion":"960010"},"reason":"Scheduled","message":"Successfully assigned ephemeral-577301787-1w711 to ip-10-0-1-88.us-west-1.compute.internal","source":{"component":"default-scheduler"},"firstTimestamp":"2017-05-02T23:27:59Z","lastTimestamp":"2017-05-02T23:27:59Z","count":1,"type":"Normal"}}
860 <24>1 2017-05-02T16:28:00-07:00 - ephemeral-577301787-1w711 - - - {"verb":"ADDED","event":{"metadata":{"name":"ephemeral-577301787-1w711.14baedca740c3745","namespace":"kube-system","selfLink":"/api/v1/namespaces/kube-system/events/ephemeral-577301787-1w711.14baedca740c3745","uid":"fb1161c5-2f8e-11e7-956e-06d49b68d344","resourceVersion":"960021","creationTimestamp":"2017-05-02T23:28:00Z"},"involvedObject":{"kind":"Pod","namespace":"kube-system","name":"ephemeral-577301787-1w711","uid":"fa78e002-2f8e-11e7-956e-06d49b68d344","apiVersion":"v1","resourceVersion":"960011","fieldPath":"spec.containers{ephemeral}"},"reason":"Pulling","message":"pulling image \"busybox\"","source":{"component":"kubelet","host":"ip-10-0-1-88.us-west-1.compute.internal"},"firstTimestamp":"2017-05-02T23:28:00Z","lastTimestamp":"2017-05-02T23:28:00Z","count":1,"type":"Normal"}}
```

kensimon commented on sinks/eventdata.go:
```diff
@@ -0,0 +1,76 @@
+		Priority:  rfc5424.Daemon,
+		Timestamp: e.Event.LastTimestamp.Time,
+		Hostname:  e.Event.ClusterName,
+		AppName:   e.Event.InvolvedObject.Name,
```
This is kinda hand-wavey... syslog's host and app fields don't have clean mappings here so I just put in some semi-useful stuff in its place to make parsers happy.  (All the information is part of the message JSON as well.)

timothysc commented on eventrouter.go:
```diff
@@ -33,8 +33,8 @@ import (
 	kubernetesWarningEventCounterVec = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "heptio_eventrouter_warnings_total",
-		Help:      "Total number of warning events in the kubernetes cluster",
+		Name: "heptio_eventrouter_warnings_total",
```
Done on other PR, so may need rebase.

timothysc commented on sinks/eventdata.go:
```diff
@@ -0,0 +1,76 @@
+	msg := rfc5424.Message{
+		Priority:  rfc5424.Daemon,
+		Timestamp: e.Event.LastTimestamp.Time,
+		Hostname:  e.Event.ClusterName,
```
It will make it look like it all came from the same source, but we do know the component names where they originated, we could decorate component.ClusterName

timothysc replied:
maybe reverse that.. cluster.component

kensimon replied:
Thinking about this a bit further, I think we can actually use a host name here... what about:

```
Hostname: e.Event.Source.Host,
AppName: e.Event.Source.Component,
```

Using a real hostname is less surprising, at least.

kensimon replied:
In practice Hostname ends up being empty (substituted with a `-`) for events that aren't scoped to a host.

timothysc replied:
> In practice Hostname ends up being empty (substituted with a -) for events that aren't scoped to a host. 

Umm that almost seems like a bug.

kensimon replied:
The rfc5424 is what's substituting a `-`, the bug is that kubernetes events bubble up without a hostname recorded.  See the comment here: https://github.com/heptio/eventrouter/pull/21#issuecomment-298964791

timothysc commented on sinks/httpsink.go:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	h.httpClient = pester.New()
+	h.httpClient.Backoff = pester.ExponentialJitterBackoff
```
Ohh man kube needs that 'pester.ExponentialJitterBackoff' in it's clients.

timothysc commented on sinks/httpsink.go:
```diff
@@ -0,0 +1,161 @@
+	for _, evt := range events {
+		w, err := evt.WriteRFC5424(h.bodyBuf)
+		written += w
+		if err != nil {
```
I have no idea what the state space of errors here could be, but bailing on getting the event out would be pretty bad.   I don't know if there are retry conditions.

kensimon replied:
Looking at the source, it happens if either the io.Writer returns an error (which bytes.Buffer is guaranteed never to do, it panics if it can't allocate enough), or if the message can't be marshalled into a byte string (which is unrecoverable, and really only happens if app names/etc are too long.  See https://github.com/crewjam/rfc5424/blob/master/marshal.go#L90)

timothysc commented on sinks/httpsink.go:
```diff
@@ -0,0 +1,161 @@
+		return
+	}
+
+	resp, err := h.httpClient.Do(req)
```
So does the client properly jitterbackoff and retry 3 times on error and only exit after that?

kensimon replied:
yep, `Do` just blocks until the library gives up.

timothysc commented on sinks/httpsink_test.go:
```diff
@@ -0,0 +1,168 @@
+	// Send the event, sleep to ensure the request is attempted
+	mockStatus = http.StatusInternalServerError
+	sink.UpdateEvents(evt, nil)
+	time.Sleep(100 * time.Millisecond)
```
We should probably have a standard like 

// TODO: SLEEP 

For trace debugging in case we hit flakes.

timothysc replied:
I can't tell you how many hours of my life I've spent digging in kube code to uncover sleep issues.

kensimon commented:
Rebased, split to two commits, addressed some of the comments.

Main thing I changed is that we're now using `event.Source.Host` for the syslog Hostname field, and `event.Source.Component` for the syslog App name field.  Host isn't set 100% of the time, but I think it's better to forward the unset value on (serialized as a `-`  in rfc5424) than try to make up something meaningful to put there.
